### PR TITLE
Do not backpropagate through layers without gradient sources

### PR DESCRIPTION
**Pull request by tbennun**
This PR tracks layers that do not need to be computed on backpropagation and skips those. Disable with `LBANN_NO_BACKPROP_DISABLE=1`.

### CODE REVIEW

File: src/models/model.cpp [hunk 1]
```diff
@@ -467,6 +467,34 @@ void model::serialize_to_onnx(onnx::ModelProto& mp)
...
+  for (int i = 0; i < l->get_num_parents(); ++i) {
+    if (m_needed_for_backprop.find(&l->get_parent_layer(i)) !=
+        m_needed_for_backprop.end()) {
+      return true;
+    }
+  }
+
+  // Second, check the layer itself. If frozen, backprop is not necessary.
+  if (l->is_frozen()) {
+    return true;
```
[the remark this answers is not in the document]
Comment (tbennun) [reply src/models/model.cpp:483]: This is a mistake, thank you for noticing!

**Comment by tbennun** [comment]
Patch is ready and working

**Review by benson31 — APPROVED** [review]
friends are bugs. but so is bad performance. :shipit:

**Review by bvanessen — APPROVED** [review]

**Comment by bvanessen** [comment]
@tbennun Does it past CI?

**Comment by tbennun** [comment]
Yes

**Pull request merged**